core-api/src/main/java/com/optimizely/ab/event/BatchEventProcessor.java

```diff
@@ -0,0 +1,204 @@
+        future = executor.submit(runnable);
+    }
+
+    @Override
+    public void close() throws Exception {
+        logger.info("Start close");
+        eventQueue.put(SHUTDOWN_SIGNAL);
+        try {
+            future.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            logger.warn("Interrupted while awaiting termination.");
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public int addHandler(NotificationHandler<LogEvent> handler) {
+        logger.debug("Adding notification handler: {}", handler);
+        return notificationManager.addHandler(handler);
+    }
+
+    public void process(UserEvent userEvent) {
+        logger.debug("Received userEvent: {}", userEvent);
+
+
+        if (executor.isShutdown()) {
+            logger.warn("Executor shutdown, not accepting tasks.");
+            return;
+        }
+
+        if (!eventQueue.offer(userEvent)) {
+            logger.warn("Payload not accepted by the queue. Current size: {}", eventQueue.size());
+        }
+    }
+
+    public class EventConsumer implements Runnable {
+        private LinkedList<UserEvent> currentBatch = new LinkedList<>();
+        private long deadline = System.currentTimeMillis() + batchDuration;
+
+        @Override
+        public void run() {
+            try {
+                while (true) {
+                    if (System.currentTimeMillis() > deadline) {
+                        logger.debug("Deadline exceeded flushing current batch.");
+                        flush();
+                    }
+
+                    Object item = eventQueue.poll(50, TimeUnit.MILLISECONDS);
+                    if (item == null) {
+                        logger.info("Empty item, sleeping for 50ms.");
+                        Thread.sleep(50);
+                        continue;
+                    }
+
+                    if (item == SHUTDOWN_SIGNAL) {
+                        logger.info("Received shutdown signal.");
+                        break;
+                    }
+
+                    addToBatch((UserEvent) item);
+                }
+            } catch (InterruptedException e) {
+                logger.info("Interrupted while processing buffer.");
+            } catch (Exception e) {
+                logger.error("Uncaught exception processing buffer.", e);
+            } finally {
+                logger.info("Exiting processing loop. Attempting to flush pending events.");
+                flush();
+            }
+        }
+
+        private void addToBatch(UserEvent userEvent) {
+            if (!currentBatch.isEmpty()) {
+                String currentRevision = currentBatch.peekLast().getUserContext().getProjectConfig().getRevision();
+
+                // Separate batches based on config revision.
+                if (!currentRevision.equals(userEvent.getUserContext().getProjectConfig().getRevision())) {
+                    flush();
+                    currentBatch = new LinkedList<>();
```

Comment:
Shouldn't there be a `currentBatch.add(userEvent)` here or better that we not return?

Reply:
Good point, and lacks coverage. Will add a test!

Reply:
Done